packages/styled-base/types/index.d.ts

```diff
@@ -26,105 +24,38 @@ export {
 
 export { ComponentSelector, Interpolation }
 
-type JSXInEl = JSX.IntrinsicElements
-
-export type WithTheme<P, T> = P extends { theme: infer Theme }
-  ? P & { theme: Exclude<Theme, undefined> }
-  : P & { theme: T }
-
 export interface StyledOptions {
   label?: string
   shouldForwardProp?(propName: string): boolean
   target?: string
 }
 
-export interface StyledComponent<InnerProps, StyleProps, Theme extends object>
-  extends React.SFC<InnerProps & StyleProps & { theme?: Theme }>,
+export interface StyledComponent<InnerProps, StyleProps>
+  extends React.SFC<InnerProps & StyleProps>,
     ComponentSelector {
   /**
    * @desc this method is type-unsafe
    */
-  withComponent<NewTag extends keyof JSXInEl>(
+  withComponent<NewTag extends keyof JSX.IntrinsicElements>(
     tag: NewTag
-  ): StyledComponent<JSXInEl[NewTag], StyleProps, Theme>
-  withComponent<Tag extends React.ComponentType<any>>(
-    tag: Tag
-  ): StyledComponent<PropsOf<Tag>, StyleProps, Theme>
+  ): StyledComponent<JSX.IntrinsicElements[NewTag], StyleProps>
+  withComponent<Props extends object>(
+    tag: React.ComponentType<Props>
+  ): StyledComponent<Props, StyleProps>
 }
 
 type ReactClassPropKeys = keyof React.ClassAttributes<any>
 
-interface CreateStyledComponentBaseThemeless<InnerProps, ExtraProps> {
-  <
-    StyleProps extends Omit<
-      Overwrapped<InnerProps, StyleProps>,
-      ReactClassPropKeys
-    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>,
-    Theme extends object = object
-  >(
-    ...styles: Array<Interpolation<WithTheme<StyleProps, Theme>>>
-  ): StyledComponent<InnerProps, StyleProps, Theme>
-  <
-    StyleProps extends Omit<
-      Overwrapped<InnerProps, StyleProps>,
-      ReactClassPropKeys
-    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>,
-    Theme extends object = object
-  >(
+export interface CreateStyledComponent<InnerProps, ExtraProps> {
+  <StyleProps extends object>(
+    ...styles: Array<Interpolation<InnerProps & StyleProps & ExtraProps>>
+  ): StyledComponent<InnerProps, StyleProps>
+  <StyleProps extends object>(
     template: TemplateStringsArray,
-    ...styles: Array<Interpolation<WithTheme<StyleProps, Theme>>>
-  ): StyledComponent<InnerProps, StyleProps, Theme>
+    ...styles: Array<Interpolation<InnerProps & StyleProps & ExtraProps>>
+  ): StyledComponent<InnerProps, StyleProps>
 }
 
-interface CreateStyledComponentBaseThemed<
```

Comment:
There is currently an issue with TS types not being tested at all by the CI (I'm going to work on fixing this soon) - but there are some existing tests around this. Could you refactor them where appropriate & also check if everything that is being testes by them is covered now? 

We'd like to ensure at least that everything can still be expressed correctly - even if with different exported types.

Reply:
Tests found an issue, have lost support for `defaultProps`, will restore this and fix up all the tests

Reply:
Will check how much putting the defaultProps support back in, seeing as that is a pattern which React is moving away from I am tempted to document a workaround by specifying the Prop types manually.

We gain improvements in type inference elsewhere, so I think it's worth the tradeoff.

To detect defaultProps we need to restore the generic type being the component, which means we cannot put constraints on the prop generic type.

So to sum up, I want to drop support for defaultProps, then put in the TypeScript docs on how to support it by using this syntax instead

`styled<PropsOf<MyComponent>>(MyComponent)`

Reply:
> To detect defaultProps we need to restore the generic type being the component, which means we cannot put constraints on the prop generic type.

Could you elaborate about those tradeoffs? Examples would be good, I'm not able to follow this PR super closely, bringing myself up to speed each time is quite time-consuming when analyzing previous & current types

In general, I would be OK with dropping `defaultProps` support out of the box if it gives us some concrete benefits. Not sure if such change is OK as part of v10 though and I wouldn't like to make any decisions about that on my own. cc @mitchellhamilton @Ailrun

Reply:
Lets take CreateStyled

```
export interface CreateStyled {
    <C extends React.ReactType<any>, ExtraProps = {}>(
        component: React.ComponentType<Props>,
        options?: StyledOptions
    ): CreateStyledComponent<PropsOf<C>, ExtraProps>

    <Tag extends keyof JSX.IntrinsicElements, ExtraProps = {}>(
        tag: Tag,
        options?: StyledOptions
    ): CreateStyledComponent<JSX.IntrinsicElements[Tag], ExtraProps>
}

vs

export interface CreateStyled {
    <Props extends object, ExtraProps = {}>(
        component: React.ComponentType<Props>,
        options?: StyledOptions
    ): CreateStyledComponent<Props, ExtraProps>

    <Tag extends keyof JSX.IntrinsicElements, ExtraProps = {}>(
        tag: Tag,
        options?: StyledOptions
    ): CreateStyledComponent<JSX.IntrinsicElements[Tag], ExtraProps>
}
```

The first the generic type is the component itself, so when we need to reference the props we need to Extract it using mapped typed + infer.

We have no generic constrains on Props in this scenario but functions like `withTheme` should.

That said, I just checked and using variant 2 in our project does not increase build times or types much at all. I might kick this discussion down the road and just support it.

Reply:
Oops, made the change to CreateStyled, not CreateThemedStyled. 

I have some compilation issues now.

For my the signature is way easier to work with. If the type inference doesn't work it's really hard to bail out. 

<img width="464" alt="image" src="https://user-images.githubusercontent.com/453152/65142122-94c3c500-da44-11e9-8b18-57eb2a3fcef2.png">

I have the default React props, but it's failed to extract the props properly. In our project I could remove a bunch of manually specified types and rely on inference with the latter. 

It also is super easy to give the user a type which allows them to grant. Let me work on going with the latter and updating the docs, then we can discuss. It *feels* like a better way to go but that's just me leaning towards simplicity

Reply:
So if you want to support defaultProps / propTypes you need to do something like this:

`styled<PropsOf<typeof ReactClassComponent0>>(ReactClassComponent0)({})`

This allows us to remove workaround functions like this monster from our codebase

```
/**
 * The `typeSafeWrappedStyledComponent` fn can be used to wrap exisitng styled components and add more props to them.
 * This would usually work natively (see https://emotion.sh/docs/typescript), however emotion doesn't handle
 * union (|) types well with props, so you can use this function to do it instead:
 *
 * typeSafeWrappedStyledComponent(MyComponent)<MyAdditionalProps>(
 *  ({ myAdditionalProp, myOrigionalProp }) => (...)
 * )
 *
 * Emotion also does not type their wrapped components to pass on props to the wrapper. This allows
 * us to do that in a typesafe manner
 */

export const typeSafeWrappedStyledComponent = <C extends WrappedComponentBase>(
    Component: C,
    options?: StyledOptions
) => {
    type Props = GetComponentProps<C>
    type ReactClassPropKeys = keyof React.ClassAttributes<any>
    type Styles<T> = Interpolation<WithTheme<Omit<T, ReactClassPropKeys>, Theme>>[]

    return <AdditionalProps extends {}>(...styles: Styles<Props & AdditionalProps>) => {
        const WrappedComponent = React.forwardRef<any, Props & AdditionalProps>((props, ref) => {
            const newProps = { ...props, ref }
            const ToRender = Component as React.FC<Props & AdditionalProps>
            return <ToRender {...newProps} />
        })

        WrappedComponent.displayName = 'StyledComponentWithMixedProps'
        return styled(WrappedComponent, options)(...styles)
    }
}
```

Reply:
Have got all the existing tests and a few new ones passing. Afaik no issues. Resolving this comment as the current signatures support refs / default props etc.